ms2query/ms2library.py

```diff
@@ -0,0 +1,332 @@
+        cosine_score, cosine_matches = map(list, zip(
+            *[x[0] for x in cosine_sim_matrix]))
+        mod_cosine_score, mod_cosine_matches = map(list, zip(
+            *[x[0] for x in mod_cosine_sim_matrix]))
+        # Transform cosine score and mod cosine matches to in between 0-1
+        normalized_cosine_matches = [1 - 0.93 ** i for i in cosine_matches]
+        normalized_mod_cos_matches = \
+            [1 - 0.93 ** i for i in mod_cosine_matches]
+
+        parent_masses = [spectrum.get("parent_mass")
+                         for spectrum in preselected_spectra_list]
+        normalized_parent_masses = [parent_mass/self.max_parent_mass
+                                    for parent_mass in parent_masses]
+
+        mass_similarity = [self.base_nr_mass_similarity **
+                           abs(spectrum.get("parent_mass") -
+                               query_spectrum.get("parent_mass"))
+                           for spectrum in preselected_spectra_list]
+
+        # Get s2v_scores
+        s2v_scores = Spec2Vec(self.s2v_model,
+                              allowed_missing_percentage=100).matrix(
+            create_spectrum_documents(preselected_spectra_list),
+            create_spectrum_documents([query_spectrum]))[:, 0]
+
+        # Get ms2ds_scores
+        query_ms2ds_embeddings = MS2DeepScore(
+            self.ms2ds_model,
+            progress_bar=False).calculate_vectors([query_spectrum])
+        preselected_ms2ds_embeddings = \
+            self.ms2ds_embeddings.loc[preselected_spectrum_ids].to_numpy()
+        ms2ds_scores = cosine_similarity_matrix(query_ms2ds_embeddings,
+                                                preselected_ms2ds_embeddings
+                                                )[0]
+
+        # Add info together into a dataframe
+        preselected_spectra_df = pd.DataFrame(
+            {"cosine_score": cosine_score,
+             "cosine_matches": normalized_cosine_matches,
+             "mod_cosine_score": mod_cosine_score,
+             "mod_cosine_matches": normalized_mod_cos_matches,
+             "parent_mass": normalized_parent_masses,
+             "mass_sim": mass_similarity,
+             "s2v_scores": s2v_scores,
+             "ms2ds_scores": ms2ds_scores},
+            index=preselected_spectrum_ids)
+        return preselected_spectra_df
+
+    @staticmethod
+    def get_ms2query_model_prediction(matches_info: Dict[str, pd.DataFrame],
+                                      ms2query_model_file_name: str
+                                      ) -> Dict[str, pd.DataFrame]:
+        """Adds ms2query predictions to dataframes
+
+        matches_info:
+            A dictionary with as keys the query spectrum ids and as values
+            pd.DataFrames containing the top 20 preselected matches and all
+            info needed about these matches to run the ms2query model.
+        ms2query_model_file_name:
+            File name of a hdf5 name containing the ms2query model.
```

Comment:
Again, for later: the model is very small. So it could easily be kept in memory, or here as `self.ms2query_nn` or similar.

Reply:
So do you mean instead of passing the filename, to pass the model that is already loaded into memory? And would you suggest the same for the ms2ds and spec2vec model?